src/tool_cb_wrt.c

```diff
@@ -137,6 +146,18 @@ size_t tool_write_cb(char *buffer, size_t sz, size_t nmemb, void *userdata)
   if(!outs->stream && !tool_create_output_file(outs))
     return failure;
 
+  if(isatty && (outs->bytes < 2000) && !config->binary_ok) {
+    /* binary output to terminal? */
+    size_t bytes = sz * nmemb;
+    if(memchr(buffer, 0, bytes)) {
+      warnf(config->global, "Binary output can mess up your terminal. "
+            "Use \"--output -\" to tell curl to output it to your terminal "
+            "anyway, or consider \"--output <FILE>\" to save to a file.\n");
+      config->synthetic_error = ERR_BINARY_TERMINAL;
+      return bytes-1;
```

Comment:
Couldn't you `return failure` here, that would be more in line with the rest of the function. In that case you don't even need the `bytes` you could just pass that multiplication directly to memchr.